### PR TITLE
Add analytic account id on hours block

### PR DESCRIPTION
**Pull request by vrenaville**
In some case invoice have mutiple invoice line with different analytic account set on it.
This improvement allow to specified the analytic account that you want to manage with this hours block, so it deduced hours only if analytic lines linked is on the account specified in the hours block.

### CODE REVIEW

**Comment by dreispt** [comment]
Could you rebase please?

**Comment by vrenaville** [comment]
It seems to have flake8 error, I will check it

File: analytic_hours_block/account_analytic.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+            context = {}
+        if 'hours_block_search_invoice_id' not in context:
+            return super(account_analytic, self).name_search(cr, uid,
+                                                             name, args,
+                                                             operator,
+                                                             context, limit)
+        else:
+            invoice_id = context['hours_block_search_invoice_id']
+            invoice_line_obj = self.pool['account.invoice.line']
+            invoice_lines_ids = invoice_line_obj.search(cr, uid,
```
Comment (dreispt) [inline analytic_hours_block/account_analytic.py]: Suggestion: if you put a line break here, the next 3 lines can become just one.

Comment (dreispt) [inline analytic_hours_block/account_analytic.py]: Suggestion: if you put a line break here, the next 4 lines can become just one.

**Comment by jgrandguillaume** [comment]
Hi Vincent,

That's a nice little improvement, thanks ! Agreed with @dreispt on his comments, a part from that, LGTM.

You have my :+1: if travis is green.

Regards,

Joël

**Comment by vrenaville** [comment]
Thanks for your Reviews !
@dreispt  I have update my code according to your comment.

Travis seems to fail on other modules :(

File: analytic_hours_block/account_analytic.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
+# -*- coding: utf-8 -*-
+from osv import orm
```
Comment (yvaucher) [inline analytic_hours_block/account_analytic.py]: please import using full path

```
from openerp.osv import orm
```

File: analytic_hours_block/account_analytic.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
+# -*- coding: utf-8 -*-
```
Comment (yvaucher) [inline analytic_hours_block/account_analytic.py:1]: License ?

File: analytic_hours_block/account_analytic.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+class account_analytic(orm.Model):
+    _inherit = 'account.analytic.account'
+
+    def name_search(self, cr, uid, name, args=None,
+                    operator='ilike', context=None, limit=80):
+        if context is None:
+            context = {}
+        if 'hours_block_search_invoice_id' not in context:
+            return super(account_analytic, self).name_search(cr, uid,
```
Comment (yvaucher) [inline analytic_hours_block/account_analytic.py]: Better to call kargs with keys `operator=operator, context=context, ...`

File: analytic_hours_block/hours_block.py [hunk 5]
```diff
@@ -31,7 +31,9 @@ def _get_last_action(self, cr, uid, ids, name, arg, context=None):
         res = {}
         for block in self.browse(cr, uid, ids, context=context):
             cr.execute("SELECT max(al.date) FROM account_analytic_line AS al"
-                       " WHERE al.invoice_id = %s", (block.invoice_id.id,))
+                       " WHERE al.invoice_id = %s "
+                       " and al.account_id = %s",
```
Comment (yvaucher) [inline analytic_hours_block/hours_block.py]: Can you uppercase `AND` ? :smile_cat:

File: analytic_hours_block/project.py [hunk 6]
```diff
@@ -11,17 +11,20 @@ def hours_block_tree_view(self, cr, uid, ids, context):
         hours_block_obj = self.pool.get('account.hours.block')
         project = self.browse(cr, uid, ids)[0]
         invoice_line_ids = invoice_line_obj.search(
-            cr, uid, [('account_analytic_id', '=', project.analytic_account_id.id)])
+            cr, uid,
+            [('account_analytic_id',
+              '=',
+              project.analytic_account_id.id)])
```
Comment (yvaucher) [inline analytic_hours_block/project.py]: missing context propagation

File: analytic_hours_block/project.py [hunk 7]
```diff
@@ -11,17 +11,20 @@ def hours_block_tree_view(self, cr, uid, ids, context):
...
         invoice_lines = invoice_line_obj.browse(cr, uid, invoice_line_ids)
         invoice_ids = [x.invoice_id.id for x in invoice_lines]
-        res_ids = hours_block_obj.search(
-            cr, uid, [('invoice_id', 'in', invoice_ids)])
+        res_ids = hours_block_obj.search(cr, uid, [('invoice_id',
+                                                    'in', invoice_ids)])
```
Comment (yvaucher) [inline analytic_hours_block/project.py]: another context missing

**Comment by yvaucher** [comment]
@vrenaville I just added some nitpicking otherwise :+1:

File: analytic_hours_block/account_analytic.py [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+
+class account_analytic(orm.Model):
+    _inherit = 'account.analytic.account'
+
+    def name_search(self, cr, uid, name, args=None,
+                    operator='ilike', context=None, limit=80):
+        if context is None:
+            context = {}
+        if 'hours_block_search_invoice_id' not in context:
```
Comment (lepistone) [inline analytic_hours_block/account_analytic.py]: maybe it is more readable if you switch `if` and `else` and remove `not`?

File: analytic_hours_block/hours_block.py [hunk 9]
```diff
@@ -163,8 +169,8 @@ def _get_invoice(self, cr, uid, ids, context=None):
         block_ids = set()
         inv_obj = self.pool.get('account.invoice')
         for invoice in inv_obj.browse(cr, uid, ids, context=context):
-            block_ids.update(
-                [inv.id for inv in invoice.account_hours_block_ids])
+            block_ids.update([inv.id for inv
+                              in invoice.account_hours_block_ids])
```
Comment (lepistone) [inline analytic_hours_block/hours_block.py:173]: I wouldn't split the line here. Wasn't it OK before?

File: analytic_hours_block/project.py (hunk 7 above)
Comment (lepistone) [inline analytic_hours_block/project.py]: Also here, I do not see an improvement in the split.

**Comment by lepistone** [comment]
I also :+1: with minor remarks. Thanks

File: analytic_hours_block/hours_block.py (hunk 9 above)
Comment (vrenaville) [inline analytic_hours_block/hours_block.py:173]: It's just for pep8, it's worked before but failed pep8 check

**Comment by vrenaville** [comment]
@lepistone @yvaucher 
Thanks for your review and comment !
 I have update my proposal according to your comments

File: analytic_hours_block/account_analytic.py (hunk 3 above)
Comment (yvaucher) [inline analytic_hours_block/account_analytic.py:1]: Ok thanks (fixed)

**Comment by yvaucher** [comment]
:+1:

**Comment by lepistone** [comment]
Ok, I checked that the red build is unrelated. Thanks!

**Pull request merged**